heclib/heclib_c/src/Public/zpdRetrieve.c

```diff
@@ -705,8 +706,272 @@ int zpdRetrieve(long long *ifltab, zStructPairedData *pds, int retrieveSizeFlag)
+						"Record has no conversion information.\n"
+						"Values not converted.",
+						pds->locationStruct->verticalDatum == 0 ? "UNSET"   :
+						pds->locationStruct->verticalDatum == 1 ? "NAVD-88" :
+						pds->locationStruct->verticalDatum == 2 ? "NGVD-29" : "OTHER",
+						cvertical_datum);
+					zmessage(ifltab, errmsg);
+				}
+				else {
+					//---------------------------------------------------------------//
+					// ensure the vertical datum info is returned in the user header //
+					//---------------------------------------------------------------//
+					verticalDatumInfoToString(&vdiStr, vdi, TRUE);
+					char *headerString;
+					if (pds->userHeader) {
+						headerString = userHeaderToString(pds->userHeader, pds->userHeaderNumber);
+						if (!strstr(headerString, VERTICAL_DATUM_INFO_USER_HEADER_PARAM)) {
+							headerString = (char *)realloc(
+								headerString,
+								strlen(headerString)
+								+ VERTICAL_DATUM_INFO_USER_HEADER_PARAM_LEN
+								+ strlen(vdiStr)
+								+ 3);
+							sprintf(
+								headerString+strlen(headerString), 
+								"%s:%s;", 
+								VERTICAL_DATUM_INFO_USER_HEADER_PARAM,
+								vdiStr);
+							free(pds->userHeader);
+						}
+					}
+					else {
+						headerString = (char *)malloc(
+							VERTICAL_DATUM_INFO_USER_HEADER_PARAM_LEN
+							+ strlen(vdiStr)
+							+ 3);
+						sprintf(
+							headerString, 
+							"%s:%s;", 
+							VERTICAL_DATUM_INFO_USER_HEADER_PARAM,
+							vdiStr);
+					}
+					free(vdiStr);
+					//--------------------------------------------//
+					// add the requested datum to the user header //
+					//--------------------------------------------//
+					headerString = (char *)realloc(
+						headerString,
+						strlen(headerString)
+						+ VERTICAL_DATUM_USER_HEADER_PARAM_LEN
+						+ strlen(cvertical_datum)
+						+ 3);
+					sprintf(
+						headerString+strlen(headerString),
+						"%s:%s;", 
+						VERTICAL_DATUM_USER_HEADER_PARAM,
+						cvertical_datum);
+					pds->userHeader = stringToUserHeader(headerString, &pds->userHeaderNumber);
+					pds->allocated[zSTRUCT_userHeader] = TRUE;
+					free(headerString);
+					if (indElev) {
+						//-----------------------------//	
+						// determine the offset to use //
+						//-----------------------------//	
+						double offset;
+						switch(ivertical_datum) {
+							case IVERTICAL_DATUM_NAVD88 :
+								offset = vdi->offsetToNavd88;
+								break;
+							case IVERTICAL_DATUM_NGVD29 :
+								offset = vdi->offsetToNgvd29;
+								break;
+							default :
+								if(!strcmp(cvertical_datum, vdi->nativeDatum) || !strcmp(cvertical_datum, CVERTICAL_DATUM_OTHER)) {
+									offset = 0;
+								}
+								else {
+									offset = UNDEFINED_VERTICAL_DATUM_VALUE;
+								}
+								break;
+						}
+						if (offset != 0.) {
+							offset = getOffset(offset, vdi->unit, pds->unitsIndependent);
+							if (offset == UNDEFINED_VERTICAL_DATUM_VALUE) {
```

Comment:
The following if block is redundant. We are already in a block where indElev is TRUE.